modules/channel-telegram/src/backend/client.ts

```diff
@@ -0,0 +1,139 @@
+    order: 100
+  })
+
+  async function outgoingHandler(event: sdk.IO.Event, next: sdk.IO.MiddlewareNextCallback) {
+    if (event.channel !== 'telegram') {
+      return next()
+    }
+
+    const client = clients[event.botId]
+    if (!client) {
+      return next()
+    }
+
+    const messageType = event.type === 'default' ? 'text' : event.type
+    const chatId = event.threadId || event.target
+
+    if (!_.includes(outgoingTypes, messageType)) {
+      return next(new Error('Unsupported event type: ' + event.type))
+    }
+
+    if (messageType === 'typing') {
+      const typing = parseTyping(event.payload.value)
+      await client.telegram.sendChatAction(chatId, 'typing')
+      await Promise.delay(typing)
+    } else if (messageType === 'text') {
+      // Quick Replies
+      const keyboard = Markup.keyboard(keyboardButtons<Button>(event.payload.quick_replies))
+
+      if (event.payload.markdown != false) {
+        // Attempt at sending with markdown first, fallback to regular text on failure
+        await client.telegram
+          .sendMessage(chatId, event.preview, Extra.markdown(true).markup({ ...keyboard, one_time_keyboard: true }))
+          .catch(() =>
+            client.telegram.sendMessage(
+              chatId,
+              event.preview,
+              Extra.markdown(false).markup({ ...keyboard, one_time_keyboard: true })
+            )
+          )
+      } else {
+        await client.telegram.sendMessage(
+          chatId,
+          event.preview,
+          Extra.markdown(false).markup({ ...keyboard, one_time_keyboard: true })
+        )
+      }
+    } else if (messageType === 'carousel') {
+      if (event.payload.elements && event.payload.elements.length) {
+        const { title, picture, subtitle } = event.payload.elements[0]
+        const buttons = event.payload.elements.map(x => x.buttons)
+
+        if (picture) {
+          await client.telegram.sendChatAction(chatId, 'upload_photo')
+          await client.telegram.sendPhoto(chatId, { url: picture, filename: path.basename(picture) })
+        }
+        const keyboard = keyboardButtons<CallbackButton>(buttons)
+        await client.telegram.sendMessage(
+          chatId,
+          `*${title}*\n${subtitle}`,
+          Extra.markdown(true).markup(Markup.inlineKeyboard(keyboard))
+        )
+      }
+    } else {
+      throw new Error(`Message type "${messageType}" not implemented yet`)
+    }
+
+    next(undefined, false)
+  }
+}
+
+function parseTyping(typing) {
+  if (isNaN(typing)) {
+    return 1000
+  }
+
+  return Math.max(typing, 500)
+}
+
+function keyboardButtons<T>(arr: any[] | undefined): T[] | undefined {
+  if (!arr || !arr.length) {
+    return undefined
+  }
+
+  const rows = arr[0].length ? arr : [arr]
+
+  return rows.map(
+    row =>
+      row.map(x => {
+        if (x.url) {
+          return Markup.urlButton(x.title, x.url)
+        }
+
+        return Markup.callbackButton(x.title, x.payload)
+      }) as any
+  )
```

Comment:
I have a general comment on this file, since you don't implement all telegram features, can you add some comments on where some features can be added so the community can take if from there easily. e.g : support other content-types, keyboard button mappings, etc.